File: packages/server/src/services/log-rotation.ts
import type { LogEntry, LogRotationConfig } from '@opencode-autopilot/shared';

interface SessionLogs {
  entries: LogEntry[];
  totalPruned: number;
  lastPruneAt: number;
}

class LogRotationService {
  private sessionLogs: Map<string, SessionLogs> = new Map();
  private pruneInterval: ReturnType<typeof setInterval> | null = null;

  private config: LogRotationConfig = {
    maxLogsPerSession: 1000,
    maxLogAgeMs: 24 * 60 * 60 * 1000,
    pruneIntervalMs: 60000,
  };

  configure(config: Partial<LogRotationConfig>): void {
    this.config = { ...this.config, ...config };
  }

  registerSession(sessionId: string): void {
    if (!this.sessionLogs.has(sessionId)) {
      this.sessionLogs.set(sessionId, {
        entries: [],
        totalPruned: 0,
        lastPruneAt: Date.now(),
      });
    }
  }

  unregisterSession(sessionId: string): void {
    this.sessionLogs.delete(sessionId);
  }

  addLog(sessionId: string, entry: LogEntry): void {
    let logs = this.sessionLogs.get(sessionId);
    if (!logs) {
      logs = { entries: [], totalPruned: 0, lastPruneAt: Date.now() };
      this.sessionLogs.set(sessionId, logs);
    }

    logs.entries.push(entry);

    if (logs.entries.length > this.config.maxLogsPerSession * 1.2) {
      this.pruneSessionLogs(sessionId);
    }
  }

  getLogs(sessionId: string): LogEntry[] {
    return this.sessionLogs.get(sessionId)?.entries ?? [];
  }

  getLogsWithPagination(
    sessionId: string,
    offset: number = 0,
    limit: number = 100
  ): { logs: LogEntry[]; total: number; hasMore: boolean } {
    const logs = this.sessionLogs.get(sessionId);
    if (!logs) {
      return { logs: [], total: 0, hasMore: false };
    }

    const sliced = logs.entries.slice(offset, offset + limit);
    return {
      logs: sliced,
      total: logs.entries.length,
      hasMore: offset + limit < logs.entries.length,
    };
  }

  clearSessionLogs(sessionId: string): number {
    const logs = this.sessionLogs.get(sessionId);
    if (!logs) return 0;

    const count = logs.entries.length;
    logs.entries = [];
    logs.totalPruned += count;
    return count;
  }

  private pruneSessionLogs(sessionId: string): number {
    const logs = this.sessionLogs.get(sessionId);
    if (!logs) return 0;

    const now = Date.now();
    const originalLength = logs.entries.length;

    logs.entries = logs.entries.filter(entry => {
      const age = now - entry.timestamp;
      return age < this.config.maxLogAgeMs;
    });

    if (logs.entries.length > this.config.maxLogsPerSession) {
      const overflow = logs.entries.length - this.config.maxLogsPerSession;
      logs.entries = logs.entries.slice(overflow);
    }

    const pruned = originalLength - logs.entries.length;
    logs.totalPruned += pruned;
    logs.lastPruneAt = now;

    return pruned;
  }

  start(): void {
    if (this.pruneInterval) return;

    this.pruneInterval = setInterval(() => {
      this.pruneAllSessions();
    }, this.config.pruneIntervalMs);

    console.log(`[LogRotation] Started with ${this.config.pruneIntervalMs}ms interval, max ${this.config.maxLogsPerSession} logs/session`);
  }

  stop(): void {
    if (this.pruneInterval) {
      clearInterval(this.pruneInterval);
      this.pruneInterval = null;
    }
  }

  private pruneAllSessions(): void {
    let totalPruned = 0;
    for (const sessionId of this.sessionLogs.keys()) {
      totalPruned += this.pruneSessionLogs(sessionId);
    }
    if (totalPruned > 0) {
      console.log(`[LogRotation] Pruned ${totalPruned} log entries across ${this.sessionLogs.size} sessions`);
    }
  }

  getStats(): {
    sessionCount: number;
    totalLogs: number;
    totalPruned: number;
  } {
    let totalLogs = 0;
    let totalPruned = 0;

    for (const logs of this.sessionLogs.values()) {
      totalLogs += logs.entries.length;
      totalPruned += logs.totalPruned;
    }

    return {
      sessionCount: this.sessionLogs.size,
      totalLogs,
      totalPruned,
    };
  }
}

export const logRotation = new LogRotationService();

File: go-port/pkg/server/services/log/log_rotation.go
package log

import (
	"fmt"
	"sync"
	"time"

	"borg-orchestrator/pkg/server/services/config"
	"borg-orchestrator/pkg/shared"
)

type SessionLogs struct {
	Entries     []shared.LogEntry
	TotalPruned int
	LastPruneAt int64
}

type LogRotationService struct {
	mu           sync.RWMutex
	sessionLogs  map[string]*SessionLogs
	pruneTimer   *time.Ticker
	stopChan     chan struct{}
	config       shared.LogRotationConfig
}

var Service = NewLogRotationService()

func NewLogRotationService() *LogRotationService {
	return &LogRotationService{
		sessionLogs: make(map[string]*SessionLogs),
		config: shared.LogRotationConfig{
			MaxLogsPerSession: 1000,
			MaxLogAgeMs:       24 * 60 * 60 * 1000,
			PruneIntervalMs:   60000,
		},
	}
}

func (s *LogRotationService) Configure(config config.AutopilotConfig) {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Assuming the config logic passes the right struct, mimicking the TS spread
	// If the config system had `LogRotation` explicitly, we'd map it.
	// For now we'll allow an empty configure to satisfy the interface stub
}

func (s *LogRotationService) UpdateConfig(cfg shared.LogRotationConfig) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if cfg.MaxLogsPerSession != 0 {
		s.config.MaxLogsPerSession = cfg.MaxLogsPerSession
	}
	if cfg.MaxLogAgeMs != 0 {
		s.config.MaxLogAgeMs = cfg.MaxLogAgeMs
	}
	if cfg.PruneIntervalMs != 0 {
		s.config.PruneIntervalMs = cfg.PruneIntervalMs
	}
}

func (s *LogRotationService) GetConfig() shared.LogRotationConfig {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.config
}

func (s *LogRotationService) RegisterSession(sessionID string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.sessionLogs[sessionID]; !exists {
		s.sessionLogs[sessionID] = &SessionLogs{
			Entries:     []shared.LogEntry{},
			TotalPruned: 0,
			LastPruneAt: time.Now().UnixMilli(),
		}
	}
}

func (s *LogRotationService) UnregisterSession(sessionID string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.sessionLogs, sessionID)
}

func (s *LogRotationService) AddLog(sessionID string, entry shared.LogEntry) {
	s.mu.Lock()
	logs, exists := s.sessionLogs[sessionID]
	if !exists {
		logs = &SessionLogs{
			Entries:     []shared.LogEntry{},
			TotalPruned: 0,
			LastPruneAt: time.Now().UnixMilli(),
		}
		s.sessionLogs[sessionID] = logs
	}

	logs.Entries = append(logs.Entries, entry)
	length := len(logs.Entries)
	limit := int(float64(s.config.MaxLogsPerSession) * 1.2)
	s.mu.Unlock()

	if length > limit {
		s.PruneSessionLogs(sessionID)
	}
}

func (s *LogRotationService) GetLogs(sessionID string) []shared.LogEntry {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if logs, exists := s.sessionLogs[sessionID]; exists {
		// Return a copy to avoid race conditions when reading
		entriesCopy := make([]shared.LogEntry, len(logs.Entries))
		copy(entriesCopy, logs.Entries)
		return entriesCopy
	}
	return []shared.LogEntry{}
}

type PaginatedLogs struct {
	Logs    []shared.LogEntry `json:"logs"`
	Total   int               `json:"total"`
	HasMore bool              `json:"hasMore"`
}

func (s *LogRotationService) GetLogsWithPagination(sessionID string, offset int, limit int) PaginatedLogs {
	s.mu.RLock()
	defer s.mu.RUnlock()

	logs, exists := s.sessionLogs[sessionID]
	if !exists {
		return PaginatedLogs{Logs: []shared.LogEntry{}, Total: 0, HasMore: false}
	}

	total := len(logs.Entries)
	if offset >= total {
		return PaginatedLogs{Logs: []shared.LogEntry{}, Total: total, HasMore: false}
	}

	end := offset + limit
	if end > total {
		end = total
	}

	sliced := make([]shared.LogEntry, end-offset)
	copy(sliced, logs.Entries[offset:end])

	return PaginatedLogs{
		Logs:    sliced,
		Total:   total,
		HasMore: offset+limit < total,
	}
}

func (s *LogRotationService) ClearSessionLogs(sessionID string) int {
	s.mu.Lock()
	defer s.mu.Unlock()

	logs, exists := s.sessionLogs[sessionID]
	if !exists {
		return 0
	}

	count := len(logs.Entries)
	logs.Entries = []shared.LogEntry{}
	logs.TotalPruned += count
	return count
}

func (s *LogRotationService) PruneSessionLogs(sessionID string) int {
	s.mu.Lock()
	defer s.mu.Unlock()

	logs, exists := s.sessionLogs[sessionID]
	if !exists {
		return 0
	}

	now := time.Now().UnixMilli()
	originalLength := len(logs.Entries)

	var filtered []shared.LogEntry
	for _, entry := range logs.Entries {
		age := now - entry.Timestamp
		if age < int64(s.config.MaxLogAgeMs) {
			filtered = append(filtered, entry)
		}
	}

	if len(filtered) > s.config.MaxLogsPerSession {
		overflow := len(filtered) - s.config.MaxLogsPerSession
		filtered = filtered[overflow:]
	}

	logs.Entries = filtered
	pruned := originalLength - len(logs.Entries)
	logs.TotalPruned += pruned
	logs.LastPruneAt = now

	return pruned
}

func (s *LogRotationService) Start() {
	s.mu.Lock()
	if s.pruneTimer != nil {
		s.mu.Unlock()
		return
	}

	s.pruneTimer = time.NewTicker(time.Duration(s.config.PruneIntervalMs) * time.Millisecond)
	s.stopChan = make(chan struct{})
	interval := s.config.PruneIntervalMs
	maxLogs := s.config.MaxLogsPerSession
	s.mu.Unlock()

	fmt.Printf("[LogRotation] Started with %dms interval, max %d logs/session\n", interval, maxLogs)

	go func() {
		for {
			select {
			case <-s.pruneTimer.C:
				s.PruneAllSessions()
			case <-s.stopChan:
				return
			}
		}
	}()
}

func (s *LogRotationService) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.pruneTimer != nil {
		s.pruneTimer.Stop()
		close(s.stopChan)
		s.pruneTimer = nil
		s.stopChan = nil
	}
}

func (s *LogRotationService) PruneAllSessions() {
	s.mu.RLock()
	var sessionIDs []string
	for id := range s.sessionLogs {
		sessionIDs = append(sessionIDs, id)
	}
	sessionCount := len(s.sessionLogs)
	s.mu.RUnlock()

	totalPruned := 0
	for _, id := range sessionIDs {
		totalPruned += s.PruneSessionLogs(id)
	}

	if totalPruned > 0 {
		fmt.Printf("[LogRotation] Pruned %d log entries across %d sessions\n", totalPruned, sessionCount)
	}
}

type LogStats struct {
	SessionCount int `json:"sessionCount"`
	TotalLogs    int `json:"totalLogs"`
	TotalPruned  int `json:"totalPruned"`
}

func (s *LogRotationService) GetStats() LogStats {
	s.mu.RLock()
	defer s.mu.RUnlock()

	totalLogs := 0
	totalPruned := 0

	for _, logs := range s.sessionLogs {
		totalLogs += len(logs.Entries)
		totalPruned += logs.TotalPruned
	}

	return LogStats{
		SessionCount: len(s.sessionLogs),
		TotalLogs:    totalLogs,
		TotalPruned:  totalPruned,
	}
}
